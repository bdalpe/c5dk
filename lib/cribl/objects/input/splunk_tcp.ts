import {Input, InputProps} from "./index";
import {Construct} from "constructs";

export interface SplunkTCPInputProps extends InputProps {
	maxS2SVersion?: string;
	useFwdTimezone?: boolean;
}

export class SplunkTCPInput extends Input {
	constructor(scope: Construct, id: string, props: SplunkTCPInputProps) {
		super(scope, id, props);
	}

	get kind() {
		return 'splunk'
	}
}
