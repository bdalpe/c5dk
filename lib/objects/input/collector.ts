import {Input} from "./index";

export class AzureBlobCollectorInput extends Input {
	get kind() {
		return 'azure_blob_collector'
	}
}

export class CriblLakeCollectorInput extends Input {
	get kind() {
		return 'cribl_lake_collector'
	}
}

export class DatabaseCollectorInput extends Input {
	get kind() {
		return 'database_collector'
	}
}

export class GoogleCloudStorageCollectorInput extends Input {
	get kind() {
		return 'google_cloud_storage_collector'
	}
}

export class HealthCheckCollectorInput extends Input {
	get kind() {
		return 'health_check_collector'
	}
}

export class RESTCollectorInput extends Input {
	get kind() {
		return 'rest_collector'
	}
}

export class S3CollectorInput extends Input {
	get kind() {
		return 's3_collector'
	}
}

export class SplunkSearchCollectorInput extends Input {
	get kind() {
		return 'splunk_search_collector'
	}
}
