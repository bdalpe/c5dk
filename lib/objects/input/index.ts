import Registry from "../../registry";
import {ConfigConstruct, ConstructProps} from "../../config";
import {Construct} from "constructs";
import {SecureVersion} from "node:tls";

interface MetadataField {
	/**
	 * Field name
	 */
	name: string;

	/**
	 * JavaScript expression to compute field's value, enclosed in quotes or backticks. (Can evaluate to a constant.)
	 */
	value: string;
}

export type InputProps = {
	/**
	 * Enable/disable this input
	 */
	disabled: boolean;

	/**
	 * Select whether to send data to Routes, or directly to Destinations.
	 */
	sendToRoutes: boolean;

	/**
	 * Pipeline to process data from this Source before sending it through the Routes.
	 */
	pipeline?: string;

	/**
	 * Fields to add to events from this input.
	 */
	metadata?: MetadataField[];

	/**
	 * Optionally, enable this config only on a specified Git branch. If empty, will be enabled everywhere.
	 */
	environment?: string;

	/**
	 * Add tags for filtering and grouping in Stream.
	 */
	streamtags?: []
}

export type NetworkInputProps = InputProps & PersistentQueueProps & {
	/**
	 * Address to bind on.
	 *
	 * Defaults to 0.0.0.0 (all addresses).
	 *
	 * @optional
	 */
	host?: string;

	/**
	 * Port to listen to.
	 */
	port: number;

	tls?: TLSProps;

	/**
	 * Regex matching IP addresses that are allowed to send data
	 */
	ipWhitelistRegex?: string;

	/**
	 * Enable if the connection is proxied by a device that supports Proxy Protocol V1 or V2.
	 */
	enableProxyHeader: boolean;
}

export type BaseTCPInputProps = NetworkInputProps & {
	/**
	 * Enable if the connection is proxied by a device that supports Proxy Protocol V1 or V2.
	 */
	enableProxyHeader: boolean;
}

export type TCPInputProps = BaseTCPInputProps & {
	/**
	 * Maximum number of active connections allowed per Worker Process, use 0 for unlimited.
	 */
	maxActiveCxn: number;
}

export type HTTPInputProps = BaseTCPInputProps & {
	/**
	 * Maximum number of active requests per Worker Process. Use 0 for unlimited.
	 */
	maxActiveReq: number;

	/**
	 * Toggle this to `true` to add request headers to events, in the `__headers` field.
	 */
	captureHeaders: boolean;

	/**
	 * How often request activity is logged at the `info` level. A value of 1 would log every request, 10 every 10th request, etc.
	 */
	activityLogSampleRate: number;

	/**
	 * How long to wait for an incoming request to complete before aborting it. Use 0 to disable.
	 */
	requestTimeout: number;
}

export interface BreakerProps {
	/**
	 * A list of event breaking rulesets that will be applied, in order, to the input data stream.
	 */
	breakerRulesets: string[];

	/**
	 * The amount of time (in milliseconds) the Event Breaker will wait for new data to be sent to a specific channel, before flushing the data stream out, as-is, to the Pipelines.
	 */
	staleChannelFlushMs: number;
}

export interface PersistentQueueEnabledInputProps {
	pqEnabled: true;
	pq: PersistentQueueSettings;
}

export interface PersistentQueueDisabledInputProps {
	pqEnabled: false;
	pq?: PersistentQueueSettings;
}

export type PersistentQueueProps = PersistentQueueDisabledInputProps | PersistentQueueEnabledInputProps;

export interface QuickConnectSettings {
	sendToRoutes: true;
	connections?: QuickConnectConnection[];
}

export interface DataRouteSettings {
	sendToRoutes: false;
	connections: QuickConnectConnection[];
}

export type QuickConnectProps = QuickConnectSettings | DataRouteSettings;

export interface PersistentQueueSettings {
	/**
	 * With Smart mode, PQ will write events to the filesystem only when it detects backpressure from the processing engine.
	 *
	 * With Always On mode, PQ will always write events directly to the queue before forwarding them to the processing engine.
	 */
	mode: 'smart' | 'always';

	/**
	 * The maximum amount of events to hold in-memory before dumping the events to disk.
	 */
	maxBufferSize: number;

	/**
	 * The number of events to send downstream before committing that Stream has read them.
	 */
	commitFrequency: number;

	/**
	 * The maximum size to store in each queue file before closing and optionally compressing (KB, MB, etc.).
	 */
	maxFileSize: string;

	/**
	 * The maximum amount of disk space the queue is allowed to consume.
	 *
	 * Once reached, the system stops queueing and applies the fallback Queue-full behavior.
	 *
	 * Enter a numeral with units of KB, MB, etc.
	 */
	maxSize: string;

	/**
	 * The location for the persistent queue files.
	 *
	 * To this field's value, the system will append: /<worker-id>/inputs/<input-id>.
	 */
	path: string;

	/**
	 * Codec to use to compress the persisted data.
	 */
	compress: 'none' | 'gzip';
}

export type TLSProps = {disabled: false} & TLSSettings | {disabled: true} & Partial<TLSSettings>;

export interface TLSSettings {
	/**
	 * The name of the predefined certificate.
	 */
	certificateName: string;

	/**
	 * Path on server containing the private key to use. PEM format. Can reference $ENV_VARS.
	 */
	privKeyPath: string;

	/**
	 * Passphrase to use to decrypt private key.
	 */
	passphrase?: string;

	/**
	 * Path on server containing certificates to use. PEM format. Can reference $ENV_VARS.
	 */
	certPath: string;

	/**
	 * Path on server containing CA certificates to use. PEM format. Can reference $ENV_VARS.
	 */
	caPath?: string;

	/**
	 * Whether to require clients to present their certificates. Used to perform client authentication using SSL certs.
	 */
	requestCert?: boolean;

	/**
	 * Minimum TLS version to accept from connections.
	 */
	minVersion?: SecureVersion;

	/**
	 * Maximum TLS version to accept from connections.
	 */
	maxVersion?: SecureVersion;
}

export interface QuickConnectConnection {
	pipeline: string;
	output: string;
}

export abstract class Input<T extends ConstructProps> extends ConfigConstruct {
	constructor(scope: Construct, id: string, props: T) {
		super(scope, id, props);
	}

	static get defaults(): Partial<InputProps> {
		return {
			disabled: false,
			sendToRoutes: true,
			streamtags: [],
		};
	}

	get kind() {
		return 'inputs';
	}

	get config() {
		return {
			type: this.kind,
			...super.config
		}
	}
}

export abstract class PQInput<T extends ConstructProps> extends Input<T> {
	static get defaults(): Partial<InputProps & PersistentQueueInputProps> {
		return {
			...super.defaults,
			pqEnabled: false,
		}
	}
}

export abstract class HTTPInput<T extends HTTPInputProps> extends Input<T> {
	static get defaults(): Partial<HTTPInputProps> {
		return {
			...super.defaults,
			maxActiveReq: 256,
			captureHeaders: false,
			activityLogSampleRate: 1,
			requestTimeout: 0,
			enableProxyHeader: false,
		}
	}
}

// Register Input class in the registry
Registry.register('inputs', Input);
