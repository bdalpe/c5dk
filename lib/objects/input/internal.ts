import {
	BreakerProps, HTTPInput,
	HTTPInputProps,
	Input,
	InputProps,
	PersistentQueueProps, QuickConnectProps,
	TCPInputProps
} from "./index";

export type CriblLogsInputProps = InputProps & QuickConnectProps & PersistentQueueProps;

export class CriblLogsInput extends Input<CriblLogsInputProps> {
	get kind() {
		return 'cribl';
	}
}

export type CriblMetricsInputProps = InputProps & QuickConnectProps & PersistentQueueProps & {
	/**
	 * A prefix that is applied to the metrics provided by Cribl Stream.
	 */
	prefix: string;

	/**
	 * Include granular metrics.
	 *
	 * Disabling this will drop the following metrics events:
	 * * `cribl.logstream.host.(in_bytes,in_events,out_bytes,out_events)`,
	 * * `cribl.logstream.index.(in_bytes,in_events,out_bytes,out_events)`,
	 * * `cribl.logstream.source.(in_bytes,in_events,out_bytes,out_events)`,
	 * * `cribl.logstream.sourcetype.(in_bytes,in_events,out_bytes,out_events)`.
	 */
	fullFidelity: boolean;
}

export class CriblMetricsInput extends Input<CriblMetricsInputProps> {
	get kind() {
		return 'criblmetrics';
	}
}

export type CriblHTTPInputProps = HTTPInputProps & {
	/**
	 * Shared secrets to be provided by any client (Authorization: <token>).
	 *
	 * If empty, unauthed access is permitted.
	 */
	authTokens: string[];
}

export class CriblHTTPInput extends HTTPInput<CriblHTTPInputProps> {
	static get defaults(): Partial<HTTPInputProps> {
		return {
			...super.defaults,
			enableProxyHeader: false,
		}
	}

	get kind() {
		return 'cribl_http';
	}
}

export class CriblTCPInput extends Input<Omit<TCPInputProps, 'ipWhitelistRegex'>> {
	get kind() {
		return 'cribl_tcp';
	}
}

export type AppScopeInputProps = TCPInputProps & BreakerProps;

export class AppScopeInput extends Input<AppScopeInputProps> {
	get kind() {
		return 'appscope';
	}
}

export type DatagenInputProps = InputProps & PersistentQueueProps & {
	/**
	 * List of datagens
	 */
	samples: {
		/**
		 * Name of the datagen file
		 */
		sample: string;

		/**
		 *  Maximum no. of events to generate per second per worker node. Defaults to 10.
		 */
		eventsPerSec: number;
	}[];
}

export class DatagenInput extends Input<DatagenInputProps> {
	get kind() {
		return 'datagen';
	}
}

export type SystemMetricsInputProps = InputProps;

export class SystemMetricsInput extends Input<SystemMetricsInputProps> {
	get kind() {
		return 'systemmetrics';
	}
}

export type SystemStateInputProps = InputProps;

export class SystemStateInput extends Input<SystemStateInputProps> {
	get kind() {
		return 'systemstate';
	}
}

export type ExecInputProps = InputProps;

export class ExecInput extends Input<ExecInputProps> {
	get kind() {
		return 'exec';
	}
}

export type FileMonitorInputProps = InputProps;

export class FileMonitorInput extends Input<FileMonitorInputProps> {
	get kind() {
		return 'filemonitor';
	}
}

export type JournalFilesInputProps = InputProps & {
	path: string;
	interval: number;
	journals: string[];
	rules: {
		filter: string;
		descriptions: string;
	}[];
	currentBoot: boolean;
	maxAgeDur: string;
};

export class JournalFilesInput extends Input<JournalFilesInputProps> {
	get kind() {
		return 'journal_files';
	}
}

// Note: Only the parent Input class is registered in the registry
