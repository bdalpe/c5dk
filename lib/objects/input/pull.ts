import {Input} from "./index";

export class SplunkSearchInput extends Input {
	get kind() {
		return 'splunk_search';
	}
}

export class AmazonMSKInput extends Input {
	get kind() {
		return 'amazon_msk';
	}
}

export class AmazonKinesisInput extends Input {
	get kind() {
		return 'amazon_kinesis';
	}
}

export class AmazonSQSInput extends Input {
	get kind() {
		return 'amazon_sqs';
	}
}

export class AmazonS3Input extends Input {
	get kind() {
		return 'amazon_s3';
	}
}

export class AzureEventHubInput extends Input {
	get kind() {
		return 'azure_event_hub';
	}
}

export class AzureBlobStorageInput extends Input {
	get kind() {
		return 'azure_blob_storage';
	}
}

export class GoogleCloudPubSubInput extends Input {
	get kind() {
		return 'google_cloud_pub_sub';
	}
}

export class Office365Input extends Input {
	get kind() {
		return 'office365';
	}
}

export class Office365ActivityInput extends Input {
	get kind() {
		return 'office365_activity';
	}
}

export class Office365MessageTraceInput extends Input {
	get kind() {
		return 'office365_message_trace';
	}
}

export class PrometheusScraperInput extends Input {
	get kind() {
		return 'prometheus_scraper';
	}
}

export class KafkaInput extends Input {
	get kind() {
		return 'kafka';
	}
}

export class ConfluentCloudInput extends Input {
	get kind() {
		return 'confluent_cloud';
	}
}

export class CrowdStrikeFDRInput extends Input {
	get kind() {
		return 'crowdstrike_fdr';
	}
}

export class WizInput extends Input {
	get kind() {
		return 'wiz';
	}
}
