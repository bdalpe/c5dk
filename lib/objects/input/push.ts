import {Input} from "./index";

export class TCPJSONInput extends Input {
	get kind() {
		return 'tcpjson';
	}
}

export class SplunkHECInput extends Input {
	get kind() {
		return 'splunk_hec';
	}
}

export class FirehoseInput extends Input {
	get kind() {
		return 'firehose';
	}
}

export class PrometheusRemoteWriteInput extends Input {
	get kind() {
		return 'prometheus_remote_write';
	}
}

export class GrafanaInput extends Input {
	get kind() {
		return 'grafana';
	}
}

export class LokiInput extends Input {
	get kind() {
		return 'loki';
	}
}

export class HTTPInput extends Input {
	get kind() {
		return 'http';
	}
}

export class RawHTTPIOInput extends Input {
	get kind() {
		return 'raw_http';
	}
}

export class ElasticSearchInput extends Input {
	get kind() {
		return 'elasticsearch';
	}
}

export class MetricsInput extends Input {
	get kind() {
		return 'metrics';
	}
}

export class SNMPTrapInput extends Input {
	get kind() {
		return 'snmptrap';
	}
}

export class TCPInput extends Input {
	get kind() {
		return 'tcp';
	}
}

export class RawUDPInput extends Input {
	get kind() {
		return 'raw_udp';
	}
}

export class NetFlowInput extends Input {
	get kind() {
		return 'netflow';
	}
}

export class OpenTelemetryInput extends Input {
	get kind() {
		return 'opentelemetry';
	}
}

export class ModelDrivenTelemetryInput extends Input {
	get kind() {
		return 'mdt';
	}
}

export class DatadogAgentInput extends Input {
	get kind() {
		return 'datadog';
	}
}

export class WindowsEventForwarderInput extends Input {
	get kind() {
		return 'wef';
	}
}
