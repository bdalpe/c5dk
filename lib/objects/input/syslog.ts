import {Input, InputProps} from "./index";
import {Construct} from "constructs";

export interface SyslogInputProps extends InputProps {}

export class SyslogInput extends Input {
	constructor(scope: Construct, id: string, props: SyslogInputProps) {
		super(scope, id, props);
	}

	get kind() {
		return 'syslog'
	}
}
