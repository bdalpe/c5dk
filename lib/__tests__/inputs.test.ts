import {it, expect, vi, beforeEach, describe} from 'vitest';
import {vol} from 'memfs'
import {Cribl} from "../index";
import {CriblHTTPInput, CriblTCPInput} from "../objects/input/internal";

vi.mock('node:fs');
vi.mock('node:fs/promises');

let cribl: Cribl;

beforeEach(() => {
	// reset the state of in-memory fs
	vol.reset()

	cribl = new Cribl({outdir: '/tmp'});
});

describe('Inputs', () => {
	it('should register the inputs', () => {
		new CriblHTTPInput(cribl, 'cribl_http', {
			mocked: true,
			config: {
				value: 'string',
				array: [
					1,
					'2',
					{value: 3}
				]
			}
		});
		new CriblTCPInput(cribl, 'cribl_tcp', {
			mocked: true,
		});
		cribl.synth();

		const volume = vol.toJSON();
		expect(Object.keys(volume)).toEqual(expect.arrayContaining(['/tmp/local/cribl/inputs.yml']));
		expect(volume['/tmp/local/cribl/inputs.yml']).toMatchSnapshot();
	})
})
