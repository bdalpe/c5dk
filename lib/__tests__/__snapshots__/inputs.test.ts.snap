// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Inputs > should register the inputs 1`] = `
"inputs:
  - cribl_http:
      type: cribl_http
      mocked: true
      config:
        value: string
        array:
          - 1
          - '2'
          - value: 3
  - cribl_tcp:
      type: cribl_tcp
      mocked: true
"
`;
